sdks/go/pkg/beam/testing/teststream/teststream.go

```diff
@@ -0,0 +1,150 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package teststream contains code configuring the TestStream primitive for
+// use in testing code that is meant to be run on streaming data sources.
+//
+// See https://beam.apache.org/blog/test-stream/ for more information.
+//
+// TestStream is supported on the Flink runner.
+package teststream
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/mtime"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/protox"
+
+	pipepb "github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1"
+)
+
+const urn = "beam:transform:teststream:v1"
+
+// Config holds information used to create a TestStreamPayload object.
+type Config struct {
+	elmType   beam.FullType
+	events    []*pipepb.TestStreamPayload_Event
+	endpoint  *pipepb.ApiServiceDescriptor
+	watermark int64
+}
+
+// NewConfig returns a Config to build a sequence of a test stream's events.
+// Requires that users provide the coder for the elements they are trying to emit.
+func NewConfig() Config {
+	return Config{elmType: nil,
+		events:    []*pipepb.TestStreamPayload_Event{},
+		endpoint:  &pipepb.ApiServiceDescriptor{},
+		watermark: 0,
```

Comment:
Nit: should probably start the watermark at the [mtime.MinTimestamp](https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/graph/mtime/time.go#L32)

Reply:
Done.